lib/routes/oauth.js

```diff
@@ -0,0 +1,187 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+var validators = require('./validators')
+var HEX_STRING = validators.HEX_STRING
+var BASE64_JWT = validators.BASE64_JWT
+
+var butil = require('../crypto/butil')
+
+var URL = require('url')
+var Request = require('request')
+var jwtool = require('fxa-jwtool')
+
+
+module.exports = function (
+  log,
+  P,
+  isA,
+  error,
+  signer,
+  config
+  ) {
+
+  var secretKey = jwtool.JWK.fromFile(config.secretKeyFile)
+  var publicKey = jwtool.JWK.fromFile(config.publicKeyFile)
+
+  var oauthProtocol = config.oauth.use_insecure_connection ? 'http' : 'https'
+  var oauthHostname = config.oauth.host
+  if (config.oauth.use_insecure_connection) {
+    if (config.oauth.port !== '80') {
+      oauthHostname += ':' + config.oauth.port
+    }
+  } else {
+    if (config.oauth.port !== '443') {
+      oauthHostname += ':' + config.oauth.port
+    }
+  }
+
+  function proxyToOAuthServer(req, reply) {
+    var url = JSON.parse(JSON.stringify(req.url))
```

Comment:
Yeah, it's my I-forget-whether-we-have-an-object-clone-function-available-in-this-scope fallback for quick hacks